### PR TITLE
perf(glm5next): compact MTP prefill outputs before MoE

### PR DESCRIPTION
**Pull request by voipmonitor**
## Purpose and status

Status: **implemented and qualified** for GLM-5.3 MTP prefill.

The GLM-5.3 MTP attention layer must process every prompt token because it
owns draft MLA KV and sparse-index state. Only one tail row per request feeds
the first draft proposal, but the existing forward also sends every prompt row
through the MTP MoE and shared output head.

This pull request keeps the full attention input, then compacts the residual
stream to the request-tail rows before post-attention normalization, MoE and
the shared output head. The MTP speculator consumes the already compact rows
and clears the selection after prefill. Capability detection keeps every other
autoregressive draft model on its existing full-row contract.

## Correctness contract

- MTP attention always receives all scheduled prompt rows.
- Draft MLA KV and sparse-index caches are populated exactly as before.
- Only rows selected by `last_token_indices` enter the prefill MoE and logits
  path.
- Multi-step decode clears the selector and retains its existing row layout.

## Performance evidence

Hardware was four stock-clock NVIDIA RTX PRO 6000 Blackwell Workstation
Edition GPUs (physical devices 4-7), TP4/DCP1, B12X target attention, B12X
NVFP4 W4A4 MoE, B12X linear kernels and PCIe all-reduce, FlashKDA prefill,
MTP3 with B12X attention and Humming MoE, full CUDA graphs, FP8 target KV,
512-token target/recurrent cache blocks and `max_num_batched_tokens=4096`.

An exact 32,770-token cold completion prompt was repeated three times per arm:

| MTP prefill output | Median TTFT | Prefill throughput |
|---|---:|---:|
| Full prompt rows through MTP MoE | 2.30193 s | 14,235.9 tok/s |
| Request-tail rows through MTP MoE | 2.24851 s | 14,574.1 tok/s |

Selective output saved 53.4 ms and improved throughput by 2.38%. A 30-second
CC1 decode check retained approximately 90 target steps/s; the optimization is
disabled after prefill and is not intended to alter decode throughput.

The complete TP4/DCP4 full-CKV stack measured 12,846 tok/s median across three
32k runs with the selective path enabled.

## Model-output validation

Six deterministic serving cases covering short prompts, an unaligned long
prompt and multiple generated tokens selected identical token IDs and text
between full-row and selective execution. The maximum selected-token
log-probability delta was 0.01357. Repeating the full-row reference against
itself produced a larger maximum delta of 0.02853, so the observed floating
point movement is within ordinary repeated-run drift for this stack.

The same six cases selected identical tokens between DCP1 and TP4/DCP4
full-CKV execution; the maximum selected-token log-probability delta was
0.00896.

## Tests

```text
.venv/bin/python -m pytest -q \
  tests/v1/core/test_mamba_align_chunk_split.py \
  tests/models/test_glm5next_model.py \
  tests/v1/core/test_kv_cache_utils.py
181 passed
```

All pre-commit hooks for the changed files pass, including Ruff and mypy. The
model unit test verifies that compact execution is numerically equal to
selecting the same rows from full execution and that the MTP MoE receives only
the selected row count.

## Relationship to open work

This pull request is stacked on the GLM-5.3 target-GDN checkpoint branch. The
checkpoint commit controls scheduler boundaries; this pull request removes
discarded MTP MoE work inside each scheduled prefill forward. PR #530 provides
the FlashKDA target prefill path, and #517 provides DCP full-CKV target
selection. Neither changes MTP post-attention row selection.

No open local-inference-lab/vllm or vllm-project/vllm pull request implements
this GLM-5.3 selective post-attention MTP contract.

AI assistance was used to implement, test, benchmark, and prepare this pull
request. The submitted behavior and evidence were reviewed against the source
and runtime logs by the human submitter.

### CODE REVIEW

**Comment by voipmonitor** [comment]
@coderabbitai review

**Comment by voipmonitor** [comment]
@coderabbitai review
